src/envelope.rs

```diff
@@ -80,46 +82,99 @@ impl SentryEnvelope {
+                }
+                
+            }
         );
         match request.send_async().await {
             Ok(_) => Ok(()),
             Err(e) => Err(e.into()),
         }
     }
 
+    /**
+    * Set user IP
+    */
+    pub fn set_user_ip(&mut self, user_ip: &str) {
+        self.user_ip = user_ip.to_string();
+    }
+
     /**
      * Attempt to parse a string into an envelope
      */
-    pub fn try_new_from_body(body: String) -> Result<SentryEnvelope, AError> {
-        if body.lines().count() == 3 {
-            let header = body.lines().next().ok_or(BodyError::InvalidNumberOfLines)?;
-            let header: Value =
-                serde_json::from_str(header).map_err(|e| BodyError::InvalidHeaderJson(e))?;
-            if let Some(dsn) = header.get("dsn") {
-                if let Some(dsn_str) = dsn.as_str() {
-                    let dsn = Dsn::from_str(dsn_str)?;
-                    Ok(SentryEnvelope {
-                        dsn,
-                        raw_body: body,
-                    })
+    pub fn try_new_from_body(body: String, fullBody: Vec<u8>, is_safe: bool) -> Result<SentryEnvelope, AError> {
+        // if body.lines().count() == 3 {
+        //     let header = body.lines().next().ok_or(BodyError::InvalidNumberOfLines)?;
+        //     let header: Value =
+        //         serde_json::from_str(header).map_err(|e| BodyError::InvalidHeaderJson(e))?;
+        //     if let Some(dsn) = header.get("dsn") {
+        //         if let Some(dsn_str) = dsn.as_str() {
+        //             let dsn = Dsn::from_str(dsn_str)?;
+        //             Ok(SentryEnvelope {
+        //                 dsn,
+        //                 raw_body: fullBody,
+        //             })
+        //         } else {
+        //             Err(AError::new(BodyError::InvalidDsnValue))
+        //         }
+        //     } else {
+        //         Err(AError::new(BodyError::MissingDsnKeyInHeader))
+        //     }
+        // } else {
+        //     Err(AError::new(BodyError::InvalidNumberOfLines))
+        // }
+
+        // Parse line by line until we find the dsn (Max 10 lines)
+        let mut dsn: Option<Dsn> = None;
+        let mut lines = body.lines();
+        for _ in 0..50 {
+            let line = lines.next().ok_or(BodyError::InvalidNumberOfLines)?;
+            let header: Value = serde_json::from_str(line).map_err(|e| BodyError::InvalidHeaderJson(e))?;
+            if let Some(dsn_value) = header.get("dsn") {
+                if let Some(dsn_str) = dsn_value.as_str() {
+                    dsn = Some(Dsn::from_str(dsn_str)?);
+                    break;
                 } else {
-                    Err(AError::new(BodyError::InvalidDsnValue))
+                    return Err(AError::new(BodyError::InvalidDsnValue));
                 }
-            } else {
-                Err(AError::new(BodyError::MissingDsnKeyInHeader))
             }
```

Comment:
Don't you think that we can keep those error types as a way to provide more insights when debugging sentry tunnels ?